File: vmm/src/unsafe_read.c
// Copyright (c) 2024, Qualcomm Innovation Center, Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Clear

int unsafe_memcpy(const void *addr, unsigned long size, void *dst);

#ifndef __BINDGEN__

#include <signal.h>
#include <setjmp.h>
#include <stdint.h>
#include <string.h>

jmp_buf buf;

void signal_handler(int s)
{
	(void)s;

	longjmp(buf, 1);
}

int unsafe_memcpy(const void *addr, unsigned long size, void *dst)
{
	struct sigaction old_sigsegv, old_sigbus;
	struct sigaction sa = {
		.sa_handler = signal_handler,
	};
	int ret;

	sigemptyset(&sa.sa_mask);

	if(sigaction(SIGSEGV, &sa, &old_sigsegv))
		return -1;

	ret = sigaction(SIGBUS, &sa, &old_sigbus);
	if (ret)
		goto restore_sigsegv;

	ret = setjmp(buf);
	if (ret)
		goto out;

	memcpy(dst, addr, size);
out:
	sigaction(SIGBUS, &old_sigbus, NULL);
restore_sigsegv:
	sigaction(SIGSEGV, &old_sigsegv, NULL);
	return ret;
}

#endif

File: vmm/src/unsafe_read.rs
// Copyright (c) 2024, Qualcomm Innovation Center, Inc. All rights reserved.
// SPDX-License-Identifier: BSD-3-Clause-Clear

use std::sync::Mutex;

use anyhow::Result;
use libc::c_void;

include!(concat!(env!("OUT_DIR"), "/unsafe_read_bindings.rs"));

static LOCK: Mutex<()> = Mutex::<()>::new(());

fn cautious_memcpy_ptr(dst: *mut c_void, size: u64, src: *const c_void) -> Result<(), ()> {
    let lock = LOCK.lock().expect("Failed to lock");
    let ret = unsafe { unsafe_memcpy(src, size, dst) };
    drop(lock);
    if ret == 0 {
        Ok(())
    } else {
        Err(())
    }
}

pub fn cautious_memcpy(dst: &mut [u8], src: &[u8]) -> Result<(), ()> {
    let size = src.len().try_into().unwrap();
    cautious_memcpy_ptr(
        dst.as_mut_ptr() as *mut c_void,
        size,
        src.as_ptr() as *const c_void,
    )
}

#[cfg(test)]
mod tests {
    use std::ptr;

    use claim::assert_err;
    use libc::c_void;

    use crate::unsafe_read::cautious_memcpy_ptr;

    use super::{cautious_memcpy, unsafe_memcpy};

    #[test]
    pub fn test1() {
        let mut buf = vec![0xff; 4];
        cautious_memcpy(&mut buf, b"abcd").expect("failed to copy");
        assert_eq!(buf, [b'a', b'b', b'c', b'd']);
    }

    #[test]
    pub fn test2() {
        let mut buf = vec![0xff; 4];
        let res = cautious_memcpy_ptr(buf.as_mut_ptr() as *mut c_void, 4, ptr::null());
        assert_err!(res);
    }
}
